### PR TITLE
pd-client: provide GetTSAsync API

### PR DESCRIPTION
**Pull request by tiancaiamao**
This API can avoid creating a lot of goroutines in TiDB.

PTAL @disksing @siddontang @nolouch

### CODE REVIEW

File: pd-client/client.go [hunk 1]
```diff
@@ -430,27 +434,38 @@ var tsoReqPool = sync.Pool{
...
 }
 
-func (c *client) GetTS(ctx context.Context) (int64, int64, error) {
-	start := time.Now()
-	defer func() { cmdDuration.WithLabelValues("tso").Observe(time.Since(start).Seconds()) }()
-
+func (c *client) GetTSAsync(ctx context.Context) *tsoRequest {
 	req := tsoReqPool.Get().(*tsoRequest)
+	req.start = time.Now()
+	req.ctx = ctx
```
Comment (siddontang) [inline pd-client/client.go]: I think it is better to pass the context to Wait directly, not here.

Comment (tiancaiamao) [inline pd-client/client.go]: It should be here, you can't just consider the cancel case.
Imagine that you need pass some value to the callee, it's too late to put it in Wait.

File: pd-client/client.go [hunk 2]
```diff
@@ -35,6 +35,8 @@ type Client interface {
 	GetClusterID(ctx context.Context) uint64
 	// GetTS gets a timestamp from PD.
 	GetTS(ctx context.Context) (int64, int64, error)
+	// GetTSAsync gets a timestamp from PD, without block the caller.
+	GetTSAsync(ctx context.Context) *tsoRequest
```
Comment (coocood) [inline pd-client/client.go]: The returned type is unexported, how can outer package use it?

Comment (tiancaiamao) [inline pd-client/client.go]: This is not a good style(golint seems doesn't complain), but the code actually works.
I just keep the minimal changes, should I rename it to uppercase? @coocood

Comment (coocood) [inline pd-client/client.go]: I think defining an exported interface with `Wait` method is better.

Comment (coocood) [inline pd-client/client.go]: Like `TSFuture`

Comment (tiancaiamao) [inline pd-client/client.go]: Why an interface rather than a struct?

Comment (coocood) [inline pd-client/client.go]: To hide the implementation detail.
`tsoRequest` is not something the user need to know.

File: pd-client/client.go [hunk 3]
```diff
@@ -445,6 +445,10 @@ func (c *client) GetTSAsync(ctx context.Context) *tsoRequest {
 	return req
 }
 
+type TSFuture interface {
+	Wait() (int64, int64, error)
```
Comment (coocood) [inline pd-client/client.go]: Add comment

**Comment by coocood** [comment]
LGTM

**Pull request merged**